diff --git a/init/Android.bp b/init/Android.bp
index 3287331e2e79894990b963cabe5819008c0dff79..ea24a5b442060b665e1a3a92cc55e92db99ec5d6 100644
--- a/init/Android.bp
+++ b/init/Android.bp
@@ -27,10 +27,10 @@ cc_defaults {
         "-Wno-unused-parameter",
         "-Werror",
         "-DALLOW_LOCAL_PROP_OVERRIDE=1",
-        "-DALLOW_PERMISSIVE_SELINUX=0",
-        "-DREBOOT_BOOTLOADER_ON_PANIC=0",
+        "-DALLOW_PERMISSIVE_SELINUX=1",
+        "-DREBOOT_BOOTLOADER_ON_PANIC=1",
         "-DWORLD_WRITABLE_KMSG=0",
-        "-DDUMP_ON_UMOUNT_FAILURE=0",
+        "-DDUMP_ON_UMOUNT_FAILURE=1",
         "-DSHUTDOWN_ZERO_TIMEOUT=0",
     ],
     product_variables: {
diff --git a/init/Android.mk b/init/Android.mk
index 2295bd719cbe32429b7a78f6018283fb9a51a32b..679e96ed98053a0e6fbd5f2ea258b15345ef43ac 100644
--- a/init/Android.mk
+++ b/init/Android.mk
@@ -20,6 +20,16 @@ init_options += \
     -DDUMP_ON_UMOUNT_FAILURE=0
 endif
 
+ifeq ($(TARGET_DEVICE),mozart)
+init_options += \
+    -UALLOW_PERMISSIVE_SELINUX \
+    -DALLOW_PERMISSIVE_SELINUX=1 \
+    -UREBOOT_BOOTLOADER_ON_PANIC \
+    -DREBOOT_BOOTLOADER_ON_PANIC=1 \
+    -UDUMP_ON_UMOUNT_FAILURE \
+    -DDUMP_ON_UMOUNT_FAILURE=1
+endif
+
 ifneq (,$(filter eng,$(TARGET_BUILD_VARIANT)))
 init_options += \
     -DSHUTDOWN_ZERO_TIMEOUT=1
diff --git a/mkbootimg/mkbootimg b/mkbootimg/mkbootimg
index 031320d6c4ec7b3aefb8c856ed34085bc2f4c85f..2302158bd1993ae93e4a233c47e110a70aeb8e41 100755
--- a/mkbootimg/mkbootimg
+++ b/mkbootimg/mkbootimg
@@ -19,6 +19,7 @@ from argparse import ArgumentParser, FileType, Action
 from os import fstat
 from struct import pack
 from hashlib import sha1
+import ctypes
 import sys
 import re
 
@@ -71,7 +72,7 @@ def write_header(args):
         args.base + args.ramdisk_offset,                # physical load addr
         filesize(args.second),                          # size in bytes
         args.base + args.second_offset,                 # physical load addr
-        args.base + args.tags_offset,                   # physical addr for kernel tags
+        ctypes.c_uint32(args.base + args.tags_offset).value, # physical addr for kernel tags
         args.pagesize,                                  # flash page size we assume
         max(args.header_version, filesize(args.dt)),    # version of bootimage header or dt size in bytes
         (args.os_version << 11) | args.os_patch_level)) # os version and patch level
